From 34169fcfa46879278eb5d06f4e035d875db2f604 Mon Sep 17 00:00:00 2001
From: Pradeep M <pradeep.m@intrinsyc.com>
Date: Tue, 22 Jan 2019 19:29:20 +0530
Subject: [PATCH 06/10] [8397] Add OSD JD9161 panel support & enable splash
 screen

Port Panel support from android lk

To enable splash screen, use below commands for dsi & hdmi

fastboot oem select-display-panel dsi
fastboot oem select-display-panel hdmi

Change-Id: I5d3d1165361f78dc7c3c0a17713eba93a1d1913f
(cherry picked from commit 1a6863aac1f853a1eb7b1da40e109e478a870b20)
---
 .../include/panel_jd9161_fwvga_video.h        | 311 ++++++++++++++++++
 platform/msm8996/gpio.c                       |   5 +
 platform/msm8996/include/platform/gpio.h      |   1 +
 platform/msm_shared/include/regulator.h       |   7 +
 target/msm8996/include/target/display.h       |   2 +-
 target/msm8996/oem_panel.c                    |  41 ++-
 target/msm8996/regulator.c                    |  23 ++
 target/msm8996/rules.mk                       |   2 +-
 target/msm8996/target_display.c               |  36 +-
 9 files changed, 401 insertions(+), 27 deletions(-)
 create mode 100644 dev/gcdb/display/include/panel_jd9161_fwvga_video.h

diff --git a/dev/gcdb/display/include/panel_jd9161_fwvga_video.h b/dev/gcdb/display/include/panel_jd9161_fwvga_video.h
new file mode 100644
index 00000000..42a95da0
--- /dev/null
+++ b/dev/gcdb/display/include/panel_jd9161_fwvga_video.h
@@ -0,0 +1,311 @@
+/* Copyright (c) 2016-2017, The Linux Foundation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ *       copyright notice, this list of conditions and the following
+ *       disclaimer in the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name of The Linux Foundation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*---------------------------------------------------------------------------
+ * This file is autogenerated file using gcdb parser. Please do not edit it.
+ * Update input XML file to add a new entry or update variable in this file
+ * VERSION = "1.0"
+ *---------------------------------------------------------------------------*/
+
+#ifndef __PANEL_JD9161_FWVGA_VIDEO_H__
+#define __PANEL_JD9161_FWVGA_VIDEO_H__
+/*---------------------------------------------------------------------------*/
+/* HEADER files                                                              */
+/*---------------------------------------------------------------------------*/
+#include "panel.h"
+
+/*---------------------------------------------------------------------------*/
+/* Panel configuration                                                       */
+/*---------------------------------------------------------------------------*/
+static struct panel_config jd9161_fwvga_video_panel_data = {
+	"qcom,mdss_dsi_jd9161_fwvga_vid", "dsi:0:", "qcom,mdss-dsi-panel",
+	10, 0, "DISPLAY_1", 0, 0, 60, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ""
+};
+
+/*---------------------------------------------------------------------------*/
+/* Panel resolution                                                          */
+/*---------------------------------------------------------------------------*/
+static struct panel_resolution jd9161_fwvga_video_panel_res = {
+	480, 854, 78, 78, 78, 0, 10, 10, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0
+};
+
+/*---------------------------------------------------------------------------*/
+/* Panel color information                                                   */
+/*---------------------------------------------------------------------------*/
+static struct color_info jd9161_fwvga_video_color = {
+	24, 0, 0xff, 0, 0, 0
+};
+
+/*---------------------------------------------------------------------------*/
+/* Panel on/off command information                                          */
+/*---------------------------------------------------------------------------*/
+static char jd9161_fwvga_video_on_cmd0[] = {
+	0x04, 0x00, 0x39, 0xC0,
+	0xBF, 0x91, 0x61, 0xF2,
+};
+
+static char jd9161_fwvga_video_on_cmd1[] = {
+	0x03, 0x00, 0x39, 0xC0,
+	0xB3, 0x00, 0x9B, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd2[] = {
+	0x03, 0x00, 0x39, 0xC0,
+	0xB4, 0x00, 0x9B, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd3[] = {
+	0x02, 0x00, 0x39, 0xC0,
+	0xC3, 0x04, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd4[] = {
+	0x07, 0x00, 0x39, 0xC0,
+	0xB8, 0x00, 0x6F, 0x01,
+	0x00, 0x6F, 0x01, 0xFF
+};
+
+static char jd9161_fwvga_video_on_cmd5[] = {
+	0x04, 0x00, 0x39, 0xC0,
+	0xBA, 0x34, 0x23, 0x00,
+};
+
+static char jd9161_fwvga_video_on_cmd6[] = {
+	0x03, 0x00, 0x39, 0xC0,
+	0xC4, 0x30, 0x6A, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd7[] = {
+	0x0A, 0x00, 0x39, 0xC0,
+	0xC7, 0x00, 0x01, 0x32,
+	0x05, 0x65, 0x2A, 0x12,
+	0xA5, 0xA5, 0XFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd8[] = {
+	0x27, 0x00, 0x39, 0xC0,
+	0xC8, 0x7F, 0x6A, 0x5A,
+	0x4E, 0x49, 0x39, 0x3B,
+	0x23, 0x37, 0x32, 0x2F,
+	0x49, 0x35, 0x3B, 0x31,
+	0x2B, 0x1E, 0x0F, 0x00,
+	0x7F, 0x6A, 0x5A, 0x4E,
+	0x49, 0x39, 0x3B, 0x23,
+	0x37, 0x32, 0x2F, 0x49,
+	0x35, 0x3B, 0x31, 0x2B,
+	0x1E, 0x0F, 0x00, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd9[] = {
+	0x11, 0x00, 0x39, 0xC0,
+	0xD4, 0x1E, 0x1F, 0x1F,
+	0x1F, 0x06, 0x04, 0x0A,
+	0x08, 0x00, 0x02, 0x1F,
+	0x1F, 0x1F, 0x1F, 0x1F,
+	0x1F, 0xFF, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd10[] = {
+	0x11, 0x00, 0x39, 0xC0,
+	0xD5, 0x1E, 0x1F, 0x1F,
+	0x1F, 0x07, 0x05, 0x0B,
+	0x09, 0x01, 0x03, 0x1F,
+	0x1F, 0x1F, 0x1F, 0x1F,
+	0x1F, 0xFF, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd11[] = {
+	0x11, 0x00, 0x39, 0xC0,
+	0xD6, 0x1F, 0x1E, 0x1F,
+	0x1F, 0x07, 0x09, 0x0B,
+	0x05, 0x03, 0x01, 0x1F,
+	0x1F, 0x1F, 0x1F, 0x1F,
+	0x1F, 0xFF, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd12[] = {
+	0x11, 0x00, 0x39, 0xC0,
+	0xD7, 0x1F, 0x1E, 0x1F,
+	0x1F, 0x06, 0x08, 0x0A,
+	0x04, 0x02, 0x00, 0x1F,
+	0x1F, 0x1F, 0x1F, 0x1F,
+	0x1F, 0xFF, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd13[] = {
+	0x15, 0x00, 0x39, 0xC0,
+	0xD8, 0x20, 0x00, 0x00,
+	0x30, 0x08, 0x20, 0x01,
+	0x02, 0x00, 0x01, 0x02,
+	0x06, 0x7B, 0x00, 0x00,
+	0x72, 0x0A, 0x0E, 0x49,
+	0x08, 0xFF, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd14[] = {
+	0x14, 0x00, 0x39, 0xC0,
+	0xD9, 0x00, 0x0A, 0x0A,
+	0x89, 0x00, 0x00, 0x06,
+	0x7B, 0x00, 0x00, 0x00,
+	0x3B, 0x33, 0x1F, 0x00,
+	0x00, 0x00, 0x03, 0x7B,
+};
+
+static char jd9161_fwvga_video_on_cmd15[] = {
+	0x01, 0x00, 0x39, 0xC0,
+	0x35, 0xFF, 0xFF, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd16[] = {
+	0x02, 0x00, 0x39, 0xC0,
+	0xBE, 0x01, 0xFF, 0xFF
+};
+
+static char jd9161_fwvga_video_on_cmd17[] = {
+	0x02, 0x00, 0x39, 0xC0,
+	0xC1, 0x10, 0xFF, 0xFF
+};
+
+static char jd9161_fwvga_video_on_cmd18[] = {
+	0x0B, 0x00, 0x39, 0xC0,
+	0xCC, 0x34, 0x20, 0x38,
+	0x60, 0x11, 0x91, 0x00,
+	0x40, 0x00, 0x00, 0xFF,
+};
+
+static char jd9161_fwvga_video_on_cmd19[] = {
+	0x02, 0x00, 0x39, 0xC0,
+	0xBE, 0X00, 0x05, 0x80,
+};
+
+static char jd9161_fwvga_video_on_cmd20[] = {
+	0x11, 0x00, 0x5, 0x80
+};
+
+static char jd9161_fwvga_video_on_cmd21[] = {
+	0x29, 0x00, 0x5, 0x80,
+};
+
+static struct mipi_dsi_cmd jd9161_wvga_video_on_command[] = {
+	{sizeof(jd9161_fwvga_video_on_cmd0), jd9161_fwvga_video_on_cmd0, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd1), jd9161_fwvga_video_on_cmd1, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd2), jd9161_fwvga_video_on_cmd2, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd3), jd9161_fwvga_video_on_cmd3, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd4), jd9161_fwvga_video_on_cmd4, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd5), jd9161_fwvga_video_on_cmd5, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd6), jd9161_fwvga_video_on_cmd6, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd7), jd9161_fwvga_video_on_cmd7, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd8), jd9161_fwvga_video_on_cmd8, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd9), jd9161_fwvga_video_on_cmd9, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd10), jd9161_fwvga_video_on_cmd10, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd11), jd9161_fwvga_video_on_cmd11, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd12), jd9161_fwvga_video_on_cmd12, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd13), jd9161_fwvga_video_on_cmd13, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd14), jd9161_fwvga_video_on_cmd14, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd15), jd9161_fwvga_video_on_cmd15, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd16), jd9161_fwvga_video_on_cmd16, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd17), jd9161_fwvga_video_on_cmd17, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd18), jd9161_fwvga_video_on_cmd18, 0x01},
+	{sizeof(jd9161_fwvga_video_on_cmd19), jd9161_fwvga_video_on_cmd19, 0x00},
+	{sizeof(jd9161_fwvga_video_on_cmd20), jd9161_fwvga_video_on_cmd20, 0x120},
+	{sizeof(jd9161_fwvga_video_on_cmd21), jd9161_fwvga_video_on_cmd21, 0x0A},
+};
+
+#define JD9161_FWVGA_VIDEO_ON_COMMAND 22
+
+static char jd9161_wvga_video_off_command[] = {
+	0x28, 0x00, 0x05, 0x80
+};
+
+
+static struct mipi_dsi_cmd jd9161_fwvga_video_off_command[] = {
+	{sizeof(jd9161_wvga_video_off_command), jd9161_wvga_video_off_command, 0x32},
+};
+
+#define JD9161_FWVGA_VIDEO_OFF_COMMAND 1
+
+
+static struct command_state jd9161_fwvga_video_state = {
+	0, 1
+};
+
+/*---------------------------------------------------------------------------*/
+/* Command mode panel information                                            */
+/*---------------------------------------------------------------------------*/
+static struct commandpanel_info jd9161_fwvga_video_command_panel = {
+	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+};
+
+/*---------------------------------------------------------------------------*/
+/* Video mode panel information                                              */
+/*---------------------------------------------------------------------------*/
+static struct videopanel_info jd9161_fwvga_video_video_panel = {
+	0, 0, 0, 0, 1, 1, DSI_NON_BURST_SYNCH_EVENT, 0, 0x9
+};
+
+/*---------------------------------------------------------------------------*/
+/* Lane configuration                                                        */
+/*---------------------------------------------------------------------------*/
+static struct lane_configuration jd9161_fwvga_video_lane_config = {
+	2, 0, 1, 1, 0, 0, 0
+};
+
+/*---------------------------------------------------------------------------*/
+/* Panel timing                                                              */
+/*---------------------------------------------------------------------------*/
+static const uint32_t jd9161_fwvga_video_timings[] = {
+	0x87, 0x1C, 0x12, 0x00, 0x40, 0x44, 0x16, 0x1E, 0x17, 0x03, 0x04, 0x00,
+};
+
+static struct panel_timing jd9161_fwvga_video_timing_info = {
+	0, 4, 0x04, 0x1d
+};
+
+/*---------------------------------------------------------------------------*/
+/* Panel reset sequence                                                      */
+/*---------------------------------------------------------------------------*/
+static struct panel_reset_sequence jd9161_fwvga_video_reset_seq = {
+	{1, 0, 1, }, {20, 2, 20, }, 2
+};
+#if 1
+/*---------------------------------------------------------------------------*/
+/* Backlight setting                                                         */
+/*---------------------------------------------------------------------------*/
+static struct backlight jd9161_fwvga_video_backlight = {
+	BL_PWM, 1, 255, 0, 2, 0
+};
+#endif
+static const uint32_t jd9161_fwvga_thulium_video_timings[] = {
+		0x1E, 0x1B, 0x04, 0x06, 0x02, 0x03, 0x04, 0xa0,
+		0x1E, 0x1B, 0x04, 0x06, 0x02, 0x03, 0x04, 0xa0,
+		0x1E, 0x1B, 0x04, 0x06, 0x02, 0x03, 0x04, 0xa0,
+		0x1E, 0x1B, 0x04, 0x06, 0x02, 0x03, 0x04, 0xa0,
+		0x1E, 0x0E, 0x04, 0x05, 0x02, 0x03, 0x04, 0xa0,
+};
+
+#endif /*__PANEL_JD9161_FWVGA_VIDEO_H_*/
diff --git a/platform/msm8996/gpio.c b/platform/msm8996/gpio.c
index ba369e40..823f5d32 100644
--- a/platform/msm8996/gpio.c
+++ b/platform/msm8996/gpio.c
@@ -52,6 +52,11 @@ void gpio_set(uint32_t gpio, uint32_t dir)
 	return;
 }
 
+uint32_t gpio_get_8996(uint32_t gpio)
+{
+	return readl((unsigned int *)GPIO_IN_OUT_ADDR(gpio));
+}
+
 /* Configure gpio for blsp uart */
 void gpio_config_uart_dm(uint8_t id)
 {
diff --git a/platform/msm8996/include/platform/gpio.h b/platform/msm8996/include/platform/gpio.h
index 9cf42f17..4fc168a6 100644
--- a/platform/msm8996/include/platform/gpio.h
+++ b/platform/msm8996/include/platform/gpio.h
@@ -60,6 +60,7 @@
 
 void gpio_config_uart_dm(uint8_t id);
 void gpio_config_blsp_i2c(uint8_t, uint8_t);
+uint32_t gpio_get_8996(uint32_t gpio);
 void gpio_set(uint32_t gpio, uint32_t dir);
 void gpio_tlmm_config(uint32_t gpio, uint8_t func, uint8_t dir, uint8_t pull,
 		uint8_t drvstr, uint32_t enable);
diff --git a/platform/msm_shared/include/regulator.h b/platform/msm_shared/include/regulator.h
index 9f02e7ef..ed79a135 100644
--- a/platform/msm_shared/include/regulator.h
+++ b/platform/msm_shared/include/regulator.h
@@ -67,11 +67,18 @@
 
 #define REG_LDO1	BIT(0)
 #define REG_LDO2	BIT(1)
+#define REG_LDO5	BIT(4)
 #define REG_LDO6	BIT(5)
+#define REG_LDO11	BIT(10)
 #define REG_LDO12	BIT(11)
 #define REG_LDO14	BIT(13)
 #define REG_LDO17	BIT(16)
+#define REG_LDO18	BIT(17)
+#define REG_LDO22	BIT(21)
 #define REG_LDO28	BIT(27)
+#define REG_SMPS3	BIT(7)
+#define REG_LDO3	BIT(2)
+#define REG_LDO4	BIT(4)
 
 void regulator_enable(uint32_t enable);
 void regulator_disable(uint32_t enable);
diff --git a/target/msm8996/include/target/display.h b/target/msm8996/include/target/display.h
index 9ee5ec3f..9233d8c5 100644
--- a/target/msm8996/include/target/display.h
+++ b/target/msm8996/include/target/display.h
@@ -78,7 +78,7 @@ static const uint32_t panel_physical_ctrl[] = { };
 #define MIPI_VSYNC_BACK_PORCH_LINES  2
 #define MIPI_VSYNC_FRONT_PORCH_LINES 4
 
-#define PWM_BL_LPG_CHAN_ID           4	/* lpg_out<3> */
+#define PWM_BL_LPG_CHAN_ID           1	/* lpg_out<0> */
 
 #define HDMI_PANEL_NAME              "hdmi"
 #define HDMI_CONTROLLER_STRING       "hdmi:"
diff --git a/target/msm8996/oem_panel.c b/target/msm8996/oem_panel.c
index 4cc81d1c..63fa598e 100644
--- a/target/msm8996/oem_panel.c
+++ b/target/msm8996/oem_panel.c
@@ -36,6 +36,7 @@
 #include <mipi_dsi.h>
 #include <qtimer.h>
 #include <platform.h>
+#include <platform/gpio.h>
 
 #include "gcdb_display.h"
 #include "include/panel.h"
@@ -60,7 +61,7 @@
 #include "include/panel_adv7533_1080p60.h"
 #include "include/panel_adv7533_720p60.h"
 #include "include/panel_hx8379a_truly_fwvga_video.h"
-
+#include "include/panel_jd9161_fwvga_video.h"
 /*---------------------------------------------------------------------------*/
 /* static panel selection variable                                           */
 /*---------------------------------------------------------------------------*/
@@ -77,6 +78,7 @@ enum {
 	ADV7533_1080P_VIDEO_PANEL,
 	ADV7533_720P_VIDEO_PANEL,
 	TRULY_FWVGA_VIDEO_PANEL,
+	JD9161_WVGA_VIDEO_PANEL,
 	UNKNOWN_PANEL
 };
 
@@ -97,6 +99,7 @@ static struct panel_list supp_panels[] = {
 	{"adv7533_1080p_video", ADV7533_1080P_VIDEO_PANEL},
 	{"adv7533_720p_video", ADV7533_720P_VIDEO_PANEL},
 	{"truly_fwvga_video", TRULY_FWVGA_VIDEO_PANEL},
+	{"jd9161_wvga_video", JD9161_WVGA_VIDEO_PANEL},
 };
 
 #define TARGET_ADV7533_MAIN_INST_0    (0x3D)
@@ -475,6 +478,32 @@ static bool init_panel_data(struct panel_struct *panelstruct,
 			jdi_qhd_dualdsi_thulium_cmd_timings,
 			MAX_TIMING_CONFIG * sizeof(uint32_t));
 		break;
+	case JD9161_WVGA_VIDEO_PANEL:
+		pan_type = PANEL_TYPE_DSI;
+		pinfo->lcd_reg_en = 1;
+		panelstruct->paneldata	  = &jd9161_fwvga_video_panel_data;
+		panelstruct->panelres	  = &jd9161_fwvga_video_panel_res;
+		panelstruct->color		  = &jd9161_fwvga_video_color;
+		panelstruct->videopanel   = &jd9161_fwvga_video_video_panel;
+		panelstruct->commandpanel = &jd9161_fwvga_video_command_panel;
+		panelstruct->state		  = &jd9161_fwvga_video_state;
+		panelstruct->laneconfig   = &jd9161_fwvga_video_lane_config;
+		panelstruct->paneltiminginfo
+					= &jd9161_fwvga_video_timing_info;
+		panelstruct->panelresetseq
+					= &jd9161_fwvga_video_reset_seq;
+		panelstruct->backlightinfo = &jd9161_fwvga_video_backlight;
+		pinfo->mipi.panel_on_cmds
+					= jd9161_wvga_video_on_command;
+		pinfo->mipi.num_of_panel_on_cmds
+					= JD9161_FWVGA_VIDEO_ON_COMMAND;
+		pinfo->mipi.panel_off_cmds
+			= jd9161_fwvga_video_off_command;
+		pinfo->mipi.num_of_panel_off_cmds
+			= JD9161_FWVGA_VIDEO_OFF_COMMAND;
+		memcpy(phy_db->timing,
+				jd9161_fwvga_thulium_video_timings, MAX_TIMING_CONFIG * sizeof(uint32_t));
+		break;
 	case JDI_QHD_DUALDSI_CMD_PANEL:
 		pan_type = PANEL_TYPE_DSI;
 		pinfo->lcd_reg_en = 1;
@@ -708,7 +737,15 @@ int oem_panel_select(const char *panel_name, struct panel_struct *panelstruct,
 		panel_id = JDI_4K_DUALDSI_VIDEO_NOFBC_PANEL;
 		break;
 	case HW_PLATFORM_DRAGON:
-		panel_id = TRULY_FWVGA_VIDEO_PANEL;
+		if (platform_is_apq8096_microsom820()) {
+			panel_id = JD9161_WVGA_VIDEO_PANEL;
+		} else {
+			gpio_tlmm_config(81, 0, GPIO_INPUT, GPIO_PULL_DOWN, GPIO_2MA, GPIO_ENABLE);
+			if (gpio_get_8996(81))
+				panel_id = JD9161_WVGA_VIDEO_PANEL;
+			else
+				panel_id = TRULY_FWVGA_VIDEO_PANEL;
+		}
 		break;
 	case HW_PLATFORM_ADP:
 		panel_id = ADV7533_720P_VIDEO_PANEL;
diff --git a/target/msm8996/regulator.c b/target/msm8996/regulator.c
index a95a42f3..d02da3b6 100644
--- a/target/msm8996/regulator.c
+++ b/target/msm8996/regulator.c
@@ -82,6 +82,23 @@ static uint32_t ldo14[][11]=
 	},
 };
 
+static uint32_t ldo22[][11]=
+{
+	{
+		LDOA_RES_TYPE, 22,
+		KEY_SOFTWARE_ENABLE, 4, GENERIC_DISABLE,
+		KEY_MICRO_VOLT, 4, 0,
+		KEY_CURRENT, 4, 0,
+	},
+
+	{
+		LDOA_RES_TYPE, 22,
+		KEY_SOFTWARE_ENABLE, 4, GENERIC_ENABLE,
+		KEY_MICRO_VOLT, 4, 3000000,
+		KEY_CURRENT, 4, 72,
+	},
+};
+
 static uint32_t ldo28[][14]=
 {
 	{
@@ -111,6 +128,9 @@ void regulator_enable(uint32_t enable)
 	if (enable & REG_LDO14)
 		rpm_send_data(&ldo14[GENERIC_ENABLE][0], 36, RPM_REQUEST_TYPE);
 
+	if (enable & REG_LDO22)
+		rpm_send_data(&ldo22[GENERIC_ENABLE][0], 36, RPM_REQUEST_TYPE);
+
 	if (enable & REG_LDO28)
 		rpm_send_data(&ldo28[GENERIC_ENABLE][0], 36, RPM_REQUEST_TYPE);
 }
@@ -126,6 +146,9 @@ void regulator_disable(uint32_t enable)
 	if (enable & REG_LDO14)
 		rpm_send_data(&ldo14[GENERIC_DISABLE][0], 36, RPM_REQUEST_TYPE);
 
+	if (enable & REG_LDO22)
+		rpm_send_data(&ldo22[GENERIC_DISABLE][0], 36, RPM_REQUEST_TYPE);
+
 	if (enable & REG_LDO28)
 		rpm_send_data(&ldo28[GENERIC_DISABLE][0], 36, RPM_REQUEST_TYPE);
 }
diff --git a/target/msm8996/rules.mk b/target/msm8996/rules.mk
index 0363607c..dfacf703 100644
--- a/target/msm8996/rules.mk
+++ b/target/msm8996/rules.mk
@@ -22,7 +22,7 @@ L2_PT_SZ     := 3
 
 DEFINES += PMI_CONFIGURED=1
 
-DEFINES += DISPLAY_SPLASH_SCREEN=0
+DEFINES += DISPLAY_SPLASH_SCREEN=1
 DEFINES += DISPLAY_TYPE_MIPI=1
 DEFINES += DISPLAY_TYPE_DSI6G=1
 
diff --git a/target/msm8996/target_display.c b/target/msm8996/target_display.c
index 69722b84..edbb7344 100644
--- a/target/msm8996/target_display.c
+++ b/target/msm8996/target_display.c
@@ -76,13 +76,13 @@
 static struct gpio_pin reset_gpio = {
   "msmgpio", 8, 3, 1, 0, 1
 };
-
+#if 0
 static struct gpio_pin lcd_reg_en = {	/* boost regulator */
   "pmi8994_gpios", 8, 3, 1, 0, 1
 };
-
+#endif
 static struct gpio_pin bklt_gpio = {	/* lcd_bklt_reg_en */
-  "pm8994_gpios", 14, 3, 1, 0, 1
+  "pm8994_gpios", 5, 3, 1, 0, 1
 };
 
 static struct gpio_pin enable_gpio = {
@@ -270,7 +270,7 @@ static int thulium_wled_backlight_ctrl(uint8_t enable)
 
 static int thulium_pwm_backlight_ctrl(uint8_t enable)
 {
-	uint8_t slave_id = 3; /* lpg at pmi */
+	uint8_t slave_id = 1; /* lpg at pm */
 
         if (enable) {
                 /* lpg channel 4 */
@@ -300,7 +300,7 @@ static int thulium_pwm_backlight_ctrl(uint8_t enable)
 		 /* DTEST4, OUT_HI */
                 pm8x41_lpg_write_sid(slave_id, PWM_BL_LPG_CHAN_ID, 0xE5, 0x01);
 		 /* LPG_ENABLE_CONTROL */
-                pm8x41_lpg_write_sid(slave_id, PWM_BL_LPG_CHAN_ID, 0x46, 0xA4);
+                pm8x41_lpg_write_sid(slave_id, PWM_BL_LPG_CHAN_ID, 0x46, 0xE4);
         } else {
 		 /* LPG_ENABLE_CONTROL */
                 pm8x41_lpg_write_sid(slave_id, PWM_BL_LPG_CHAN_ID, 0x46, 0x0);
@@ -311,39 +311,27 @@ static int thulium_pwm_backlight_ctrl(uint8_t enable)
 
 static void lcd_reg_enable(void)
 {
-	uint8_t slave_id = 2;	/* gpio at pmi */
-
-	struct pm8x41_gpio gpio = {
-                .direction = PM_GPIO_DIR_OUT,
-                .function = PM_GPIO_FUNC_HIGH,
-                .vin_sel = 2,   /* VIN_2 */
-                .output_buffer = PM_GPIO_OUT_CMOS,
-                .out_strength = PM_GPIO_OUT_DRIVE_MED,
-        };
-
-        pm8x41_gpio_config_sid(slave_id, lcd_reg_en.pin_id, &gpio);
-	pm8x41_gpio_set_sid(slave_id, lcd_reg_en.pin_id, 1);
+	gpio_tlmm_config(135, 0, GPIO_OUTPUT, PM_GPIO_PULL_UP_30, GPIO_8MA, GPIO_DISABLE);
+	gpio_set(135, 0x2);
 }
 
 static void lcd_reg_disable(void)
 {
-	uint8_t slave_id = 2;	/* gpio at pmi */
-
-	pm8x41_gpio_set_sid(slave_id, lcd_reg_en.pin_id, 0);
+	gpio_set(135, 0x0);
 }
 
 static void lcd_bklt_reg_enable(void)
 {
        struct pm8x41_gpio gpio = {
                 .direction = PM_GPIO_DIR_OUT,
-                .function = PM_GPIO_FUNC_HIGH,
+                .function = MPP_DTEST4,
                 .vin_sel = 2,   /* VIN_2 */
                 .output_buffer = PM_GPIO_OUT_CMOS,
                 .out_strength = PM_GPIO_OUT_DRIVE_LOW,
         };
 
         pm8x41_gpio_config(bklt_gpio.pin_id, &gpio);
-	pm8x41_gpio_set(bklt_gpio.pin_id, 1);
+	//pm8x41_gpio_set(bklt_gpio.pin_id, 1);
 }
 
 static void lcd_bklt_reg_disable(void)
@@ -483,6 +471,7 @@ int target_backlight_ctrl(struct backlight *bl, uint8_t enable)
 		ret = thulium_wled_backlight_ctrl(enable);
 		break;
 	case BL_PWM:
+#if 0
 		/* Enable MPP1 */
 		pmi8994_config_mpp_slave_id(PMIC_MPP_SLAVE_ID);
 	        mpp.base = PM8x41_MMP1_BASE;
@@ -494,6 +483,7 @@ int target_backlight_ctrl(struct backlight *bl, uint8_t enable)
 		} else {
 			pm8x41_enable_mpp(&mpp, MPP_DISABLE);
 		}
+#endif
 		ret = thulium_pwm_backlight_ctrl(enable);
 		break;
 	default:
@@ -674,7 +664,7 @@ static void wled_init(struct msm_panel_info *pinfo)
 
 int target_ldo_ctrl(uint8_t enable, struct msm_panel_info *pinfo)
 {
-	uint32_t val = BIT(1) | BIT(13) | BIT(27);
+	uint32_t val = BIT(1) | BIT(13) | BIT(21) |BIT(27);
 
 	if (enable) {
 		regulator_enable(val);
-- 
2.24.1

